MultiOracle.md

```diff
@@ -449,6 +449,113 @@ of `t` oracles, or else they can be set as constant parameters across all groupi
+When a contract is created with multiple oracles with varying number of digits, the contract should only define payouts for outcomes up to the maximum value that can be attested by the oracle with the minimum number of digits (thereafter referred as `min_nb_digits`).
+Recall that if this oracle signs the value `2^min_nb_digits - 1`, by convention it represents either the outcome value `2^min_nb_digits - 1` or a greater outcome.
+
+#### Equal outcomes
+
+When no difference is allowed between the oracles, one simply needs to "pre-pad" the outcome prefixes of oracles with larger number of digits with `0`s.
+For example, for the prefix `1010`, if an oracle uses `min_nb_digits + 2` digits, the additional prefix `00` should be prepended to yield `001010`.
+Dealing with outcomes that are above the maximum value that can be represented using `min_nb_digits` is [discussed later](#out-of-bound-outcomes). 
+
+#### Bounded error
+
+If differences are allowed between oracles, the [same algorithm](#Algorithm) defined for oracles with equal number of digits should be used, but it should be ran with `min_nb_digits + 1`, and setting the primary oracle to be the first one (in order of preference) in the group with `min_nb_digits`.
+For example, if the list of oracles is `[A, B, C, D]` with respective number of digits `[12, 10, 11, 10]`, the algorithm should be run with the order `[B, A, C, D]`.
+The list of prefixes should then be transformed so that:
+- the prefixes for oracles with more than `min_nb_digits + 1` should be "pre-padded" with zeros to reach the proper length,
+- the first `0` should be dropped for the prefixes for oracles with `min_nb_digits` 
+- any group of prefix that contains a prefix representing an outcome that cannot be attested to by the corresponding oracle should be discarded (e.g. in the previous example, if an outcome consists of oracle D attesting to a value greater than `2^min_nb_digits - 1` then that outcome should be discarded).
+
+#### Out of bound outcomes
+
+In the case of oracles with equal number of digits, if the event outcome is greater than what oracles can attest to, the contract can still be closed if all oracles attest to the maximum value (which is [the convention](./Oracle.md#Digit-Decomposition)).
+To enable contracts to be closed if the event outcome is greater than what can be represented using `min_nb_digits` in the case of oracles with varying number of digits, all possible outcomes above are assigned to the maximum contract payout.
+To cover all the possible outcomes, all the possible combinations of prefixes that cover the interval `[base^min_nb_digits; base^max_nb_digits]` are generated, where `max_nb_digits` if the number of digits used by the oracle(s) with greatest `nb_digits`.
+Using the previous example with the order `[A, B, C, D]` and respective number of digits `[12, 10, 11, 10]`, the list of prefixes to cover this interval would be:
+```
+1, 1111111111, 1, 1111111111
+01, 1111111111, 1, 1111111111
+```
```

Comment:
In the example above we have:` [base^min_nb_digits; base^max_nb_digits - 1] == [2^10; 2^12 - 1] == [10000000000; 111111111111]`

Is it correct to say that this interval can be represented the following 2 digit prefixes (where `_` stands for any digit): `_1_ _ _ _ _ _ _ _ _ _  ` and `1_ _ _ _ _ _ _ _ _ _ _` ?

Sorry I don't understand the notation above, is it equivalent to this?

Reply:
It's not exactly equivalent because in what your wrote the prefixes would have overlapping intervals. It would be more like `1 _ _ _ _ _ _ _ _ _ _ _` and `0 1 _ _ _ _ _ _ _ _ _ _`. Does that make sense?

Reply:
Yes that's right, thanks!

Reply:
Looking at how to interpret the notation
```
1, 1111111111, 1, 1111111111
01, 1111111111, 1, 1111111111
```
when it come to the describing list of prefixes that cover the interval `[2^10; 2^12 - 1] == [10000000000; 111111111111]`:

It seems like any of those lower level arrays of digits at index `i` should be interpreted as representing the **first digits** of a digit prefix of length `nb_digits[i]`

Padding those with `_` on the right (`_` representing any digit) to end up with a digit prefix of length `nb_digits[i]` and removing duplicates, we end up with:

```
1 _ _ _ _ _ _ _ _ _ _ _, 1111111111, 01_ _ _ _ _ _ _ _ _ _
```

Does it make sense to have `1111111111` here since `1111111111` is not contained in the `[2^10; 2^12 - 1] == [10000000000; 111111111111]` interval?

Reply:
I'm not sure I understand what you're saying, but the point here is that oracles `B` and `D` can only represent 10 digits numbers, so their maximum value is `2^10 - 1`. If the event outcome is above that, and oracles `A` and `C` attest to values greater than that, you still want to be able to close the contract, so this algorithm is covering the interval `2^10; 2^12 - 1` for oracle `A` and `2^10; 2^11 - 1` for oracle `C`, but `B` and `D` keep their max value (since anyway they can't attest to value greater than that). Does that make sense?

Reply:
Yes that part is clear thanks, what is still unclear though is why does the list of prefixes to cover this interval `[2^10; 2^12 - 1]` include `1111111111` which is outside the interval?